internal/grpc/messagesize/prometheus.go

```
@@ -0,0 +1,283 @@
+	if err != nil {
+		return nil, err
+	}
+
+	wrappedStream := newObservingClientStream(s, observer)
+	return wrappedStream, nil
+}
+
+type observingServerStream struct {
+	grpc.ServerStream
+
+	observer *messageSizeObserver
+}
+
+func newObservingServerStream(s grpc.ServerStream, observer *messageSizeObserver) grpc.ServerStream {
+	return &observingServerStream{
+		ServerStream: s,
+		observer:     observer,
+	}
+}
+
+func (s *observingServerStream) SendMsg(m any) error {
+	err := s.ServerStream.SendMsg(m)
+	if err != nil {
+		// Don't record the size of the message if there was an error sending it, since it may not have been sent.
+		return err
+	}
+
+	// Observe the size of the sent message.
+	message, ok := m.(proto.Message)
+	if !ok {
+		return nil
+	}
+
+	s.observer.Observe(message)
+	return nil
+}
+
+type observingClientStream struct {
+	grpc.ClientStream
+
+	observer *messageSizeObserver
+}
+
+func newObservingClientStream(s grpc.ClientStream, observer *messageSizeObserver) grpc.ClientStream {
+	return &observingClientStream{
+		ClientStream: s,
+		observer:     observer,
+	}
+}
+
+func (s *observingClientStream) SendMsg(m any) error {
+	err := s.ClientStream.SendMsg(m)
+	if err != nil {
+		// Don't record the size of the message if there was an error sending it, since it may not have been sent.
+		return err
+	}
+
+	// Observe the size of the sent message.
+	message, ok := m.(proto.Message)
+	if !ok {
+		return nil
+	}
+
+	s.observer.Observe(message)
+	return nil
+}
+
+func (s *observingClientStream) RecvMsg(m any) error {
+	err := s.ClientStream.RecvMsg(m)
+	if errors.Is(err, io.EOF) { // EOF indicates a successful end of stream.
+
+		// Record the total size of the messages sent during the course of the RPC call.
+		s.observer.FinishRPC()
+	}
+
+	// Note: we don't call FinishRPC() if there was a real error, since the total size of the messages sent during the
+	// course of the RPC call may not be accurate.
```

I think it's still valuable to log the size of messages received even in the event of an error. An error could just be a cancelled context. If we ignore the errored streams, I think this metric might end up misrepresenting the volume of traffic.

What do you mean by "not accurate" here?

---

Hmm, I might need to refactor the logic for this a bit. 

I was thinking that if we ever get an error back in the course of sending a message, that we don't have a definitive way of knowing whether or not anything actually went out over the wire to the recipient - all we know is that we encountered an error while doing so. 

As such, including that message in either the "individual message" Prometheus calculations or the "overall request / response size" traffic seems misleading.


---


However, if we do get a context cancelled or deadline exceeded error, then maybe we can record the previous observations anyway? 

One question that comes to mind...if we immediately cancel / exceed a deadline before the server responses, should we record that "empty" response size? That seems misleading... Maybe we should have an "error" label on the metric to split up the times for successful and failed requests?

---

Well, we'll have both client-side and server-side metrics for this, so I think it's a reasonable signal to record "bytes sent" and "bytes received", which could be different. In the case that sending returns an error, I agree that it's reasonable to not log those bytes as "sent"

---

@camdencheek 

I tweaked the total RPC size behavior to ignore errors and added a lot of tests in [`9f32121` (#55381)](https://github.com/sourcegraph/sourcegraph/pull/55381/commits/9f32121550d3f7c0b5b4c0823295787d6c1ef35d). Can you take a look at the tests to see if this is the behavior you wanted? (TestStreamingClientInterceptor in particular)

---

those tests look great! Thank you!